tutorials/18_training_interface.ipynb

```diff
@@ -303,63 +284,193 @@
+    "                 seed:int = 44):\n",
+    "        super().__init__()\n",
+    "\n",
+    "        torch.random.manual_seed(seed) #will set the seed device wide\n",
+    "        self.prior = prior\n",
+    "        self.simulator = simulator\n",
+    "\n",
+    "        self.theta = prior.sample((nsamples,))\n",
+    "        self.x = simulator(self.theta)\n",
+    "\n",
+    "    def __len__(self):\n",
+    "        return self.theta.shape[0]\n",
+    "\n",
+    "    def __getitem__(self, index:int):\n",
+    "        return self.theta[index,...], self.x[index,...]\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d8178235-291c-43e7-8376-98e42542d42d",
+   "metadata": {
+    "collapsed": false
+   },
+   "source": [
+    "We can now proceed to create a DataLoader and conduct our training loop as illustrated above."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "eb2aa734-dfca-4691-b3b5-f5be889219ba",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "train_data = NPEData(2048, prior, simulator)\n",
+    "train_loader = torch.utils.data.DataLoader(train_data, batch_size=128)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2759a702-6ce7-4e2c-8373-402fb30d7f49",
+   "metadata": {
+    "collapsed": false
+   },
+   "source": [
+    "For sake of demonstration, let's create another estimator using a masked autoregressive flow (maf). For this, we create a second dataset and use only parts of the data to construct the maf estimator."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "2e61a11a-b9d6-4211-8456-cb4ff8755d1e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sbi.neural_nets.net_builders import build_maf\n",
+    "\n",
+    "dummy_data = NPEData(64, prior, simulator, seed=43)\n",
+    "dummy_loader = torch.utils.data.DataLoader(dummy_data, batch_size=4)\n",
+    "dummy_theta, dummy_x = next(iter(dummy_loader))\n",
+    "maf_estimator = build_maf(dummy_theta, dummy_x)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7fe6624f-ce79-43dd-b661-1cced75dca09",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "optw = AdamW(list(maf_estimator.parameters()), lr=5e-4)\n",
+    "nepochs = 50\n",
+    "\n",
+    "for ep in range(nepochs):\n",
+    "    for idx, (theta_batch, x_batch) in enumerate(train_loader):\n",
+    "        optw.zero_grad()\n",
+    "        losses = maf_estimator.loss(theta_batch, condition=x_batch)\n",
+    "        loss = torch.mean(losses)\n",
+    "        loss.backward()\n",
+    "        optw.step()\n",
+    "    if ep % 10 == 0:\n",
+    "        print(\"last loss\", loss.item())"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cc63ae5f-33e2-422b-9365-f183a53c09b9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# let's compare the trained estimator to the NSF from above\n",
```

Comment:
NSF vs MAF, is this correct?

Reply:
Yes, this was the pseudo reason for trying another architecture.
The `maf` is apparently superior in this scenario. Should I point this out?

Reply:
No, that's fine. I just wanted to double check a potential typo 👍